c2/c2-client-bundle/c2-client-service/src/main/java/org/apache/nifi/c2/client/service/operation/SyncResourceOperationHandler.java

```diff
@@ -0,0 +1,132 @@
+        if (c2Client == null) {
+            throw new IllegalArgumentException("C2Client should not be null");
+        }
+        if (operandPropertiesProvider == null) {
+            throw new IllegalArgumentException("OperandPropertiesProvider should not be not null");
+        }
+        if (syncResourceStrategy == null) {
+            throw new IllegalArgumentException("Sync resource strategy should not be null");
+        }
+        if (c2Serializer == null) {
+            throw new IllegalArgumentException("C2 serializer should not be null");
+        }
+        return new SyncResourceOperationHandler(c2Client, operandPropertiesProvider, syncResourceStrategy, c2Serializer);
+    }
+
+    @Override
+    public OperationType getOperationType() {
+        return SYNC;
+    }
+
+    @Override
+    public OperandType getOperandType() {
+        return RESOURCE;
+    }
+
+    @Override
+    public Map<String, Object> getProperties() {
+        return operandPropertiesProvider.getProperties();
+    }
+
+    @Override
+    public C2OperationAck handle(C2Operation operation) {
+        String operationId = ofNullable(operation.getIdentifier()).orElse(EMPTY);
+
+        ResourcesGlobalHash resourcesGlobalHash = getOperationArg(operation, GLOBAL_HASH_FIELD, new TypeReference<>() {
+        });
+        if (resourcesGlobalHash == null) {
+            LOG.error("Resources global hash could not be constructed from C2 request");
+            return operationAck(operationId, operationState(NOT_APPLIED, "Resources global hash element was not found"));
+        }
+        List<ResourceItem> resourceItems = getOperationArg(operation, RESOURCE_LIST_FIELD, new TypeReference<>() {
+        });
+        if (resourceItems == null) {
+            LOG.error("Resource item list could not be constructed from C2 request");
+            return operationAck(operationId, operationState(NOT_APPLIED, "Resource item list element was not found"));
+        }
+
+        OperationState operationState = syncResourceStrategy.synchronizeResourceRepository(resourcesGlobalHash, resourceItems, c2Client::retrieveResourceItem,
+            relativeUrl -> c2Client.getCallbackUrl(null, relativeUrl));
+        C2OperationState resultState = operationState(
+            operationState,
+            switch (operationState) {
+                case NOT_APPLIED -> "No resource items were retrieved, please check the log for errors";
+                case PARTIALLY_APPLIED -> "Resource repository is partially synced, retrieving some items failed. Pleas check log for errors";
+                case FULLY_APPLIED -> "Agent Resource repository is in sync with the C2 server";
+                default -> "Unexpected status, please check the log for errors";
+            }
+        );
+
+        return operationAck(operationId, resultState);
+    }
+
+    private <T> T getOperationArg(C2Operation operation, String key, TypeReference<T> type) {
```

Comment:
Any way to move this to the C2OperationHandler or to an abstract class? Also changing the return type to Optional would make the caller side cleaner.

Reply:
Refactored, not sure if it became clearer